enhancements/machine-api/control-plane.md

```diff
@@ -0,0 +1,343 @@
+	// Only odd numbers are permitted.
+	Replicas *int32 `json:"replicas,omitempty"`
+
+	// EnableAutorepair defines if a Machine Health Checking resource
+	// should be created.
+	// This will autorepair faulty machine/nodes in scenarios where quorum would not be violated.
+	// e.g 1 unhealthy out of 3.
+	// Defaults to true.
+	EnableAutorepair bool `json:"enableautorepair,omitempty"`
+	
+	// ProviderSpec details Provider-specific configuration to use during node creation.
+	ProviderSpec ProviderSpec `json:"providerSpec"`
+}
+
+// ProviderSpec defines the configuration to use during node creation.
+type ProviderSpec struct {
+
+	// Value is an inlined, serialized representation of the resource
+	// configuration. It is recommended that providers maintain their own
+	// versioned API types that should be serialized/deserialized from this
+	// field.
+	Value *runtime.RawExtension `json:"value,omitempty"`
+}
+
+type ControlPlaneStatus struct {
+	// Total number of non-terminated machines targeted by this control plane
+	// (their labels match the selector).
+	// +optional
+	Replicas int32 `json:"replicas,omitempty"`
+
+	// Total number of non-terminated machines targeted by this control plane
+	// that have the desired template spec.
+	// +optional
+	UpdatedReplicas int32 `json:"updatedReplicas,omitempty"`
+
+	// Total number of fully running and ready Control Plane Machines.
+	// +optional
+	ReadyReplicas int32 `json:"readyReplicas,omitempty"`
+
+	// Total number of unavailable machines targeted by this control plane.
+	// This is the total number of machines that are still required for
+	// the deployment to have 100% available capacity. They may either
+	// be machines that are running but not yet ready or machines
+	// that still have not been created.
+	// +optional
+	UnavailableReplicas int32 `json:"unavailableReplicas,omitempty"`
+
+	// FailureReason indicates that there is a terminal problem reconciling the
+	// state, and will be set to a token value suitable for
+	// programmatic interpretation.
+	// +optional
+	FailureReason errors.ControlPlaneStatusError `json:"failureReason,omitempty"`
+
+	// ErrorMessage indicates that there is a terminal problem reconciling the
+	// state, and will be set to a descriptive error message.
+	// +optional
+	FailureMessage *string `json:"failureMessage,omitempty"`
+}
+
+type ControlPlaneStatusError string
+```
+
+### Risks and Mitigations
+
+During horizontal scaling operations there are sensitive scenarios like scaling from 1 to 2. As soon as the etcd API is notified of the new member the cluster loses quorum until that new member starts and joins the cluster. This must be still handled by the [Cluster etcd Operator](https://github.com/openshift/enhancements/blob/master/enhancements/etcd/cluster-etcd-operator.md#motivation) while the Control Plane controller should honour and short-circuit when it meets the etcd unhealthiness criteria as described in the workflows above.
```

Comment:
Currently scaling from 1 to 2 only happens during bootstrap or disaster recovery. Can you outline a scenario where we would go from 1 to 2 nodes?

Reply:
My guess is scaling from 1 to 3 entails a stop, however brief, at 2.